python/hail/history.py

```diff
@@ -0,0 +1,159 @@
+         .add_method(f.__name__, parsed_args)
+         .write(output_path))
+
+        return result
+    return decorator(_write)
+
+
+def format_args(arg, statements):
+    if isinstance(arg, list):
+        return [format_args(a, statements) for a in arg]
+    elif isinstance(arg, tuple):
+        return tuple([format_args(a, statements) for a in arg])
+    elif isinstance(arg, dict):
+        return {format_args(k, statements): format_args(v, statements) for k, v in arg.iteritems()}
+    else:
+        if isinstance(arg, HistoryMixin):
+            h = arg._history
+            statements += remove_dup_statements(statements, h.statements)
+            return h
+        else:
+            return arg
+
+
+def remove_dup_statements(statements1, statements2):
+    return [s2 for s1, s2 in itertools.izip_longest(statements1, statements2) if s2 and s1 != s2]
+
+
+class History(object):
+    def __init__(self, expr="", statements=[]):
+        self.expr = expr
+        self.statements = statements
+
+    def set_varid(self, id):
+        return History(id, self.statements + ['{} = ({})'.format(id, self.expr)])
+
+    def add_method(self, f_name, kwargs, index=None, key_name=None):
+        statements = self.statements[:]
+        f_args = ["{}={}".format(k, repr(format_args(v, statements))) for k, v in kwargs.iteritems()]
+        expr = "{expr}\n.{f_name}({f_args})".format(expr=self.expr, f_name=f_name, f_args=", ".join(f_args))
+        if index:
+            expr += "[{}]".format(index)
+        elif key_name:
+            expr += "[{}]".format(repr(key_name))
+        return History(expr, statements)
+
+    @staticmethod
+    def from_classmethod(cls_name, f_name, kwargs):
+        statements = []
+        f_args = ["{}={}".format(k, repr(format_args(v, statements))) for k, v in kwargs.iteritems()]
+        expr = "{cls_name}.{f_name}({args})".format(cls_name=cls_name, f_name=f_name, args=", ".join(f_args))
+        return History(expr, statements)
+
+    @staticmethod
+    def from_init(cls_name, kwargs):
+        statements = []
+        f_args = ["{}={}".format(k, repr(format_args(v, statements))) for k, v in kwargs.iteritems()]
+        expr = "{cls_name}({args})".format(cls_name=cls_name, args=", ".join(f_args))
+        return History(expr, statements)
+
+    def write(self, file):
+        with hadoop_write(file) as f:
+            f.write(str(self) + "\n")
+
+    def __repr__(self):
+        return self.expr
+
+    def __str__(self):
+        now = datetime.datetime.now()
+        history = "# {}\n# version: {}\n\n".format(now.isoformat(), Env._hc.version)
+        for stmt in self.statements:
+            history += (stmt + "\n\n")
+        history += ("(" + self.expr + ")")
+        try:
+            import autopep8
+            return autopep8.fix_code(history)
+        except ImportError:
+            return history
+
+
+class HistoryMixin(object):
+    def __init__(self):
+        self._history = None
+
+    def with_id(self, id):
```

Comment:
I didn't originally, but then added it back today. I'm ambivalent about it. It's nice to see exactly what the original commands were, but it doesn't add anything to the history with regards to what the user did to generate the file.